### PR TITLE
[ot_certs] Add certificate generation to ot_certs crate

### PR DESCRIPTION
**Pull request by jwnrt**
This PR builds upon https://github.com/lowRISC/opentitan/pull/20295 and adds code for generating an X.509 format certificate based on a given template file. It also adds an opentitantool command for performing the generation.

A follow-up PR will add the ability to generate C functions for the device the use to insert concrete values into variable parts of the certificate (as defined in the template).

### CODE REVIEW

File: sw/host/ot_certs/src/offsets.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/offsets.rs:85]:
```suggestion
        let Value::Variable { var: name, convert } = val else {
            panic!("add_variable can only be called on variables, not literals");
        };
```

Also, maybe an error would be better than a panic here

File: sw/host/ot_certs/src/x509.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/x509.rs:32]: Can `from_curve_name` fail if we're giving it that constant from the `openssl` crate?

If not, suggestion:

```rust
impl From<template::EcCurve> for EcGroup {
    fn from(ec_curve: template::EcCurve) -> EcGroup {
        match ec_curve {
            template::EcCurve::Prime256v1 => {
                EcGroup::from_curve_name(Nid::X9_62_PRIME256V1).unwrap()
            },
        }
    }
}
```

**Review by jwnrt — COMMENTED** [review]
Thanks @pamaury, I think generally this looks good, though I would want to do a proper pass on the generation code.

The commit to add the generator code is pretty huge, do you think there's any way we could split that up? Even into multiple PRs?

**Comment by pamaury** [comment]
The only natural split I could see is to add first the standard part of the certificate, and then the TCB extension. That would mean that the intermediate PR does not actually produce a certificate that matches the template though but I don't think that's a big issue. One could potentially split `type.rs` but it does not make sense on its own I think.

File: sw/host/ot_certs/src/x509.rs
Comment (pamaury) [reply sw/host/ot_certs/src/x509.rs:32]: It can fail from some constants that do not represent curves but I think for this one it cannot fail.

**Comment by pamaury** [comment]
I have split this PR into two: the base certificate which is this PR, and #20357 for the extensions. This should significantly reduce the size of the code to review.

File: sw/host/ot_certs/src/x509.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/x509.rs:156]: Does this need resolving?

File: sw/host/ot_certs/src/x509.rs
Comment (jwnrt) [inline sw/host/ot_certs/src/x509.rs:90]: TODO

**Review by jwnrt — COMMENTED** [review]
LGTM, I don't see any errors, and I've seen your #20326 PR so I'm sure it works

File: sw/host/opentitantool/src/command/certificate.rs
Comment (timothytrippel) [inline sw/host/opentitantool/src/command/certificate.rs:61]: Can you return a boxed struct here (like other commands: https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/opentitantool/src/command/lc.rs;drc=405fcdd37826ac2872088b9716d1e336933cfe87;l=200) so that these are printed to the console in JSON form (like all the other ottool commands) instead of just logging these?

File: sw/host/opentitantool/src/command/certificate.rs
Comment (timothytrippel) [inline sw/host/opentitantool/src/command/certificate.rs:64]: Same here, convention is to return a struct that gets serialized with any output you want to display to the console.

File: sw/host/opentitantool/src/command/certificate.rs
Comment (timothytrippel) [inline sw/host/opentitantool/src/command/certificate.rs:20]: just curious, does this just output the ASN.1 encoded template right now?

File: sw/host/ot_certs/src/x509.rs
Comment (timothytrippel) [inline sw/host/ot_certs/src/x509.rs:179]: Nit: link the spec here?

File: sw/host/ot_certs/src/x509.rs
Comment (timothytrippel) [inline sw/host/ot_certs/src/x509.rs:232]: This is not a random key is it?

File: sw/host/opentitantool/src/command/certificate.rs
Comment (jwnrt) [reply sw/host/opentitantool/src/command/certificate.rs:61]: Don't let this block the PR, but out of interest what is the use case for OTT's structured outputs?

File: sw/host/opentitantool/src/command/certificate.rs
Comment (nbdd0121) [reply sw/host/opentitantool/src/command/certificate.rs:64]: This is just a path. I think it makes sense to just write as a log.

File: sw/host/ot_certs/src/offsets.rs
Comment (nbdd0121) [inline sw/host/ot_certs/src/offsets.rs:64]: You could use the memchr crate which is already a dependency.

File: sw/host/ot_certs/src/offsets.rs
Comment (nbdd0121) [inline sw/host/ot_certs/src/offsets.rs:126]: IIUC for variables we generate random sequence and then identify offsets using these random sequence?

In that case, I think it's worth having a sanity check to ensure that there're no multiple occurrence of the subsequence just in case.

File: sw/host/ot_certs/src/types.rs
Comment (nbdd0121) [inline sw/host/ot_certs/src/types.rs:98]: Probably you could just `match (target, source)`?

File: sw/host/ot_certs/src/types.rs
Comment (nbdd0121) [inline sw/host/ot_certs/src/types.rs:112]: Do `match *source` here to save some ugly `size: *size` later.

File: sw/host/ot_certs/src/offsets.rs
Comment (nbdd0121) [inline sw/host/ot_certs/src/offsets.rs:82]: There are quite a few places where it expects just variable but no literals. It might worth adding another type
```rust
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Variable { 
    /// Name of the variable.
    var: String,
    /// Optional conversion to apply to the variable.
    convert: Option<Conversion>,
};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Value<T> {
    /// This value will be populated on the device when variables are set.
    Variable(Variable),
    /// Constant literal that will be set when the certificate is generated.
    Literal(T),
}
```

which removes a few panic sites and make things more strongly typed.

File: sw/device/silicon_creator/lib/cert/uds_creator.hjson
Comment (pamaury) [reply sw/device/silicon_creator/lib/cert/uds_creator.hjson:80]: I will leave this to another PR to keep this one small enough for review.

File: sw/host/ot_certs/src/x509.rs
Comment (pamaury) [reply sw/host/ot_certs/src/x509.rs:232]: Due to the way openssl work, we have to generate a random key (unless it's completely specified in the template with raw values for x/y), then use this value to extract the offset and clear the values.

File: sw/host/ot_certs/src/offsets.rs
Comment (pamaury) [reply sw/host/ot_certs/src/offsets.rs:126]: Yes. Yeah that makes sense.

File: sw/host/ot_certs/src/types.rs
Comment (pamaury) [reply sw/host/ot_certs/src/types.rs:98]: Good idea, thanks

File: sw/host/ot_certs/src/types.rs
Comment (pamaury) [reply sw/host/ot_certs/src/types.rs:112]: With the rewrite you suggested, I now have some matches on `VariableType::String { size }` where I don't know if this trick works?

Comment (nbdd0121) [reply sw/host/ot_certs/src/types.rs]: You can also match `&VariableType::String { size }` which gives you a `size` binding without reference.

**Comment by pamaury** [comment]
I have updated the PR, I think I addressed all issues raised. I split `cdi_owner.hjson` into cdi_0 and cdi_1. @timothytrippel let me know if I got the fields right.

File: sw/host/ot_certs/src/offsets.rs
Comment (nbdd0121) [inline sw/host/ot_certs/src/offsets.rs:121]:
```suggestion
                let Some(first_match) = finder.next() else {
                    bail!("internal error: could not find the offset of {:?} in the binary certificate, the value should be {:?}", offset, needle),
                };
```

**Review by nbdd0121 — APPROVED** [review]
LGTM apart from the nits

File: sw/host/ot_certs/src/offsets.rs
Comment (pamaury) [reply sw/host/ot_certs/src/offsets.rs:89]: When called, the variable is known to exist because one has to lookup information before, hence the panic. Also `add_variable` is not public. I would prefer to keep it this way.

File: sw/host/ot_certs/src/x509.rs
Comment (timothytrippel) [reply sw/host/ot_certs/src/x509.rs]: ah I see, thanks for clarifying. Perhaps a comment here explaining might help?

**Review by timothytrippel — APPROVED** [review]
Thanks @pamaury and @jwnrt , just a couple more cleanups on the templates, otherwise LGTM!

File: sw/device/silicon_creator/lib/cert/cdi_1.hjson
Comment (pamaury) [reply sw/device/silicon_creator/lib/cert/cdi_1.hjson:74]: Yes I think so ;)

**Pull request merged**
